### ISSUE TITLE
Classifying test instances with the SNLI model loaded from snapshot

### ISSUE THREAD

**Issue by salman1993** [pytorch/examples#120]
I train the SNLI model with the example training code and then I try to load up sample of the test dataset (~ 100 instances) and classify them using the code below. The train accuracy I get is 35%. That's too low since the best trained model gets around 78% on the validation set so sth must be wrong. I tried to classify a sample of the training set as well and the model performed poorly so that was confirmation that sth is not working. Is it sth with my code? 

Another weird thing is len(answers.vocab) returns 4 although it should be 3 since the labels are neutral, entail, contradict. This does not affect much since none of the predicted labels refer to unk which is the other extra label in answers.vocab

```python
inputs = data.Field(lower=args.lower)
answers = data.Field(sequential=False) 

test = data.TabularDataset( path=[path/to/test/data], format='json', fields={'sentence1': ('premise', inputs),
                                                                    'sentence2': ('hypothesis', inputs),
                                                                    'gold_label': ('label', answers)},
                                                            filter_pred=lambda ex: ex.label != '-' )

inputs.build_vocab(test)
inputs.vocab.vectors = torch.load([path/to/vector/cached])
answers.build_vocab(test)

# test iterator has batch size equal to length --> full test set
# # test_iter = data.BucketIterator(test, batch_size=len(test), device=args.gpu, sort_key=lambda ex: len(ex.premise) + len(ex.hypothesis))
test_iter = data.Iterator(test, batch_size=len(test), device=args.gpu)

model = torch.load([path/to/model/snapshot], map_location=lambda storage, location: storage.cuda(args.gpu))

test_full_batch =  next(iter(test_iter))  # there will be only 1 batch

predicted_scores = model(test_full_batch)
predicted_labels = torch.max(predicted_scores, 1)[1].view(test_full_batch.label.size()).data
n_correct = (predicted_labels == test_full_batch.label.data).sum()
n_total = test_full_batch.batch_size
train_acc = 100. * n_correct/n_total
print("train accuracy - %f" % train_acc)
```

**Comment by jekbradbury** [comment]
You have to use the same vocabulary you used during training. The easiest way to do that is to load the train set in your script and call inputs.build_vocab(train)

**Comment by salman1993** [comment]
I added this:
```python
train = data.TabularDataset( path=[path/to/training/data], format='json', fields={'sentence1': ('premise', inputs),
                                                                    'sentence2': ('hypothesis', inputs),
                                                                    'gold_label': ('label', answers)},
                                                            filter_pred=lambda ex: ex.label != '-' )


inputs.build_vocab(train)
inputs.vocab.vectors = torch.load([path/to/vector/cache])
answers.build_vocab(train)
```

These are the results:
```bash
$ python -i test.py
train accuracy - 48.000000
>>> len(answers.vocab)
4
>>> len(inputs.vocab)
56220
```

The test data provided here is a sample of 100 instances from the training data and the model gets much better on the training set so sth must be wrong. I am directly using the vector cache and the model from the resume snapshot that was saved during training. Thoughts?

**Comment by salman1993** [comment]
I think I found the issue:

```python
$ python -i test.py
## Building vocab on TEST data
>>> inputs.build_vocab(test)
>>> inputs.vocab.vectors = torch.load(args.vector_cache)
>>> len(inputs.vocab)
381
>>> answers.build_vocab(test)
>>> len(answers.vocab)
4
>>> vars(answers.vocab)
{'freqs': Counter({'neutral': 35, 'entailment': 33, 'contradiction': 32}), 'unk_init': 'random', 'stoi': defaultdict(<function Vocab.__init__.<locals>.<lambda> at 0x7f1b886bbb70>, {'neutral': 1, 'entailment': 2, 'contradiction': 3}), 'itos': ['<unk>', 'neutral', 'entailment', 'contradiction']}


## Building vocab on TRAIN data
>>> inputs.build_vocab(train)
>>> inputs.vocab.vectors = torch.load(args.vector_cache)
>>> len(inputs.vocab)
56220
>>> answers.build_vocab(train)
>>> len(answers.vocab)
4
>>> vars(answers.vocab)
{'freqs': Counter({'entailment': 183416, 'contradiction': 183187, 'neutral': 182764}), 'unk_init': 'random', 'stoi': defaultdict(<function Vocab.__init__.<locals>.<lambda> at 0x7f1b9971ea60>, {'entailment': 1, 'contradiction': 2, 'neutral': 3}), 'itos': ['<unk>', 'entailment', 'contradiction', 'neutral']}
```

The problem is the vocabulary mapping of the answers labels. It's different for TRAIN and TEST data. How do I get the labels from the test data but ensure the mapping that the model used during training? Is there any nice way of doing that in TorchText?

**Comment by jekbradbury** [comment]
If you do answers.build_vocab(train), the answer vocab mapping should be the same as it was during training. I’ll try to figure out what’s wrong today (and also we’ll soon have the ability to serialize torchtext Fields, which should make your use case easier)

**Comment by salman1993** [comment]
Instead of saving just the vector, I think we need to save the entire Vocab object when we are training. Still looking into it.

**Comment by salman1993** [comment]
I found the error. During training, the inputs vocabulary was built using the train,dev,test set but during testing, I was only building it using the train set. This is a good way of doing it to evaluate the model since we are building the vocabulary with words we know we will face.

But instead of just storing the word vectors that appear in my train/dev/test set, what if I wanted to build up a vocabulary with all the words that are in the glove.42b.300d.txt? Although that will take up more memory, I think that will be a good option to have since it will be useful to have for deploying a product that may contain many words that don't appear in the train/dev/test set but there are embeddings for them in the glove..txt file. Is there anyway of doing that now in TorchText?

**Comment by jekbradbury** [comment]
Yes, you can use fill_from_vocab=True.

**Comment by ebetica** [comment]
Only if you enable it https://github.com/pytorch/pytorch/blob/d602b3a834f3c6debfd5f32b435501374d98e31d/torch/nn/_functions/thnn/sparse.py#L12
